spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/PushDown.scala

```diff
@@ -0,0 +1,204 @@
+  def predicateColumnName(predicate: T) : String
+
+  private val partitionKeyColumns = tableDef.partitionKey.map(_.columnName)
+  private val clusteringColumns = tableDef.clusteringColumns.map(_.columnName)
+  private val indexedColumns = tableDef.allColumns.filter(_.isIndexedColumn).map(_.columnName)
+  private val regularColumns = tableDef.regularColumns.map(_.columnName)
+  private val allColumns = partitionKeyColumns ++ clusteringColumns ++ regularColumns
+
+  private val singleColumnPredicates = predicates.filter(isSingleColumn)
+
+  private val eqPredicates = singleColumnPredicates.filter(isEqualTo)
+  private val eqPredicatesByName = eqPredicates.groupBy(predicateColumnName)
+    .mapValues(_.take(1))       // take(1) in order not to push down more than one EQ predicate for the same column
+    .withDefaultValue(Seq.empty)
+
+  private val inPredicates = singleColumnPredicates.filter(isIn)
+  private val inPredicatesByName = inPredicates.groupBy(predicateColumnName)
+    .mapValues(_.take(1))      // take(1) in order not to push down more than one IN predicate for the same column
+    .withDefaultValue(Seq.empty)
+
+  private val rangePredicates = singleColumnPredicates.filter(isRangeComparison)
+  private val rangePredicatesByName = rangePredicates.groupBy(predicateColumnName).withDefaultValue(Seq.empty)
+
+  /** Returns the only column name referenced in the predicate */
+  private def predicateColumnName(predicate: Expression) = {
+    require(predicate.references.size == 1, "Given predicate is not a single column predicate: " + predicate)
+    predicate.references.head.name
+  }
+
+  /** Returns a first non-empty sequence. If not found, returns an empty sequence. */
+  private def firstNonEmptySeq[T](sequences: Seq[T]*): Seq[T] =
+    sequences.find(_.nonEmpty).getOrElse(Seq.empty[T])
+
+  /**
+   * Selects partition key predicates for pushdown:
+   * 1. Partition key predicates must be equality or IN predicates.
+   * 2. Only the last partition key column predicate can be an IN.
+   * 3. All partition key predicates must be used or none.
+   */
+  private val partitionKeyPredicatesToPushDown: Seq[T] = {
+    val (eqColumns, otherColumns) = partitionKeyColumns.span(eqPredicatesByName.contains)
+    val inColumns = otherColumns.headOption.toSeq.filter(inPredicatesByName.contains)
+    if (eqColumns.size + inColumns.size == partitionKeyColumns.size)
+      eqColumns.flatMap(eqPredicatesByName) ++ inColumns.flatMap(inPredicatesByName)
+    else
+      Nil
+  }
+
+  /**
+   * Selects clustering key predicates for pushdown:
+   * 1. Clustering column predicates must be equality predicates, except the last one.
+   * 2. The last predicate is allowed to be an equality or a range predicate.
+   * 3. The last predicate is allowed to be an IN predicate only if it was preceded by an equality predicate.
+   * 4. Consecutive clustering columns must be used, but, contrary to partition key, the tail can be skipped.
+   */
```

Comment:
The comment is a little bit confusing for `IN` predicate - it can be applied for the last clustering column defined in the table, not in the query - in this case all the clustering columns must be specified.